Client/Frontend/Browser/Tabs/TabGroupsManager.swift

```diff
@@ -6,137 +6,287 @@ import Foundation
+    //                    }
+    //                }
+    //            }
+    //
+    //            // 3. Url groups should have at least 2 url per search term so we remove smaller groups
+    //            let filteredGroupData = urlGroupData.filter { urlGroup in
+    //                let t = urlGroup.value
+    //                if t.count > 1 {
+    //                    return true
+    //                } else {
+    //                    if let onlyUrl = t.first, let index = urlInGroups.firstIndex(of: onlyUrl) {
+    //                        urlInGroups.remove(at: index)
+    //                    }
+    //                    return false
+    //                }
+    //            }
+    //
+    //            // 4. Filter the url list so it doesn't include same url as url groups
+    //            let filteredUrls = urlList.filter { url in
+    //                !urlInGroups.contains(url)
+    //            }
+    //
+    //            // 5. filteredGroupData contains groups of only 2 or more urls and filtered have urls that are not part of a group
+    //            completion(filteredGroupData, filteredUrls)
+    //        }
+    //    }
+
+    /// Create item groups from metadata.
+    ///
+    /// This function take a type `T` generic, and creates groups based on metadata available
+    /// from Application Services.
+    ///
     /// - Parameters:
-    ///   - profile: User Profile info
-    ///   - urlList: List of urls we want to make the groups from
-    ///   - completion: completion handler that contains [Search Term: [URL]]  dictionary and filterd URLs list that has URLs  that are not part of a group
-    static func getURLGroups(profile: Profile, urlList: [URL], completion: @escaping ([String: [URL]]?, _ filteredUrls: [URL]) -> Void) {
-        
+    ///   - profile: The user's `Profile` info
+    ///   - items: List of items we want to make the groups from. This is a generic type and
+    ///   currently only supports `Tab` and `URL`
+    ///   - ordering: Order in which we want to return groups, `.orderedAscending` or
+    ///   `.orderedDescending`. `.orderedSame` is also possible, but will return the exact
+    ///   order of the group that was provided. Note: this does not affect the groups' items,
+    ///   which will alway return in ascending order.
+    ///   - completion: completion handler that contains `[ASGroup<T>]`  dictionary and a
+    ///   filteredItems list, `[T]`, which is comprised of items from the original input
+    ///   that are not part of a group.
+    static func getGroups<T: Equatable>(with profile: Profile, from items: [T], using ordering: ComparisonResult, completion: @escaping ([ASGroup<T>]?, _ filteredItems: [T]) -> Void) {
+//        ROUX TODO: when URL support is added, fix check
+//        guard (items is [Tab] || items is [URL]) else { return completion(nil, [T]()) }
+        guard (items is [Tab]) else { return completion(nil, [T]()) }
```

Comment:
@electricRGB Does this mean its not really generic and we only accept URl or Tab? 🤔

Reply:
Yes. We're limiting it, currently to those, because that's what we have metadata on. But, due to the same implementation, essentially, we're using generics. otherwise, we have the same function, twice, which is not pretty.

Reply:
An approach here would be to keep this generic function but add two public methods that accepts [URL] and a [Tab]

What do you think?

Reply:
Sure I can do that.